base/src/main/java/com/smartdevicelink/managers/screen/choiceset/PresentKeyboardOperation.java

```diff
@@ -0,0 +1,242 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Created by brettywhite on 6/12/19 1:52 PM
+ *
+ */
+
+package com.smartdevicelink.managers.screen.choiceset;
+
+import com.smartdevicelink.managers.CompletionListener;
+import com.smartdevicelink.protocol.enums.FunctionID;
+import com.smartdevicelink.proxy.RPCNotification;
+import com.smartdevicelink.proxy.RPCResponse;
+import com.smartdevicelink.proxy.interfaces.ISdl;
+import com.smartdevicelink.proxy.rpc.KeyboardProperties;
+import com.smartdevicelink.proxy.rpc.OnKeyboardInput;
+import com.smartdevicelink.proxy.rpc.PerformInteraction;
+import com.smartdevicelink.proxy.rpc.SetGlobalProperties;
+import com.smartdevicelink.proxy.rpc.enums.InteractionMode;
+import com.smartdevicelink.proxy.rpc.enums.KeyboardEvent;
+import com.smartdevicelink.proxy.rpc.enums.LayoutMode;
+import com.smartdevicelink.proxy.rpc.listeners.OnRPCNotificationListener;
+import com.smartdevicelink.proxy.rpc.listeners.OnRPCResponseListener;
+import com.smartdevicelink.util.DebugTool;
+
+import java.lang.ref.WeakReference;
+import java.util.Collections;
+import java.util.List;
+
+class PresentKeyboardOperation implements Runnable {
+
+	private WeakReference<ISdl> internalInterface;
+	private KeyboardListener keyboardListener;
+	private KeyboardProperties originalKeyboardProperties, keyboardProperties, customConfig;
+	private boolean updatedKeyboardProperties;
+	private String initialText;
+	private OnRPCNotificationListener keyboardRPCListener;
+
+	PresentKeyboardOperation(ISdl internalInterface, KeyboardProperties originalKeyboardProperties, String initialText, KeyboardProperties customConfig, KeyboardListener keyboardListener){
+		this.internalInterface = new WeakReference<>(internalInterface);
+		this.keyboardListener = keyboardListener;
+		this.originalKeyboardProperties = originalKeyboardProperties;
+		this.keyboardProperties = originalKeyboardProperties;
+		this.customConfig = customConfig;
+		this.initialText = initialText;
+	}
+
+	@Override
+	public void run() {
+		addListeners();
+		start();
+	}
+
+	private void start(){
+		DebugTool.logInfo("Choice Operation: Executing present keyboard operation");
+		if (keyboardListener != null){
+			keyboardProperties = customConfig;
+			updatedKeyboardProperties = true;
+		}
+
+		updateKeyboardProperties(new CompletionListener() {
+			@Override
+			public void onComplete(boolean success) {
+				presentKeyboard();
+			}
+		});
+	}
+
+	// SENDING REQUESTS
+
+	private void presentKeyboard(){
+
+		if (internalInterface.get() != null){
+
+			PerformInteraction pi = getPerformInteraction();
+			pi.setOnRPCResponseListener(new OnRPCResponseListener() {
+				@Override
+				public void onResponse(int correlationId, RPCResponse response) {
+					finishOperation();
```

Comment:
This didn't check the onError method, meaning if there is an error this will never call `finishOperation`